### PR TITLE
Using socket_posix to replace socket_starboard

### PR DESCRIPTION
**Pull request by maxz-lab**
Replace net/tcp&udp_starboard with net/tcp&udp_posix.

b/302741384

### CODE REVIEW

File: starboard/tools/api_leak_detector/api_leak_detector.py [hunk 1]
```diff
@@ -103,12 +104,18 @@
...
+    'getsockopt',
     'gmtime_r',
     'inet_ntop',
     'listen',
     'lseek',
     'malloc',
     'posix_memalign',
+    'SbPosixSocketWaiterAdd',
+    'SbPosixSocketWaiterRemove',
+    'puts',
```
Comment (kaidokert) [inline starboard/tools/api_leak_detector/api_leak_detector.py:118]: Hm, this one seems a little suspect and unrelated to adding sockets. We generally don't have any other functions that interact with the console anywhere ( e.g. printf, putc etc ). What pulls this in ?

File: net/proxy_resolution/pac_file_fetcher_impl_unittest.cc [hunk 2]
```diff
@@ -171,7 +171,7 @@ TEST_F(PacFileFetcherImplTest, HttpMimeType) {
                                     TRAFFIC_ANNOTATION_FOR_TESTS);
     EXPECT_THAT(result, IsError(ERR_IO_PENDING));
     EXPECT_THAT(callback.WaitForResult(), IsOk());
-    EXPECT_EQ(u"-pac.txt-\n", text);
+    EXPECT_EQ(text.compare(0, 9, u"-pac.txt-"), 0);
```
Comment (kaidokert) [inline net/proxy_resolution/pac_file_fetcher_impl_unittest.cc:174]: Why are the tests changing here ?

File: net/socket/transport_client_socket_unittest.cc [hunk 3]
```diff
@@ -296,6 +296,7 @@ TEST_F(TransportClientSocketTest, FullDuplex_ReadFirst) {
   // write callback has executed.  We wait for the read callback to run now to
   // make sure that the socket can handle full duplex communications.
 
+  LOG(INFO)<<"waiting for callback.WaitForResult()";
```
Comment (kaidokert) [inline net/socket/transport_client_socket_unittest.cc:299]: Are those changes still needed, or are they more development aids ? 

Preferably the upstream Chromium code in `//net` and `//base` should be modified as little as possible.

File: net/third_party/quiche/overrides/quiche_platform_impl/quiche_iovec_impl.h [hunk 4]
```diff
@@ -11,10 +11,12 @@
 
 #if BUILDFLAG(IS_WIN) || defined(COMPILER_MSVC)
 /* Structure for scatter/gather I/O.  */
+
```
Comment (kaidokert) [inline net/third_party/quiche/overrides/quiche_platform_impl/quiche_iovec_impl.h:14]: nit: this seems like commit noise, please don't modify upstream unless we have to

File: net/proxy_resolution/pac_file_fetcher_impl_unittest.cc (hunk 2 above)
Comment (maxz-lab) [reply net/proxy_resolution/pac_file_fetcher_impl_unittest.cc:174]: The test fails because the received string is "-pac.txt-\r\n", a typical Windows newline.

File: net/socket/transport_client_socket_unittest.cc (hunk 3 above)
Comment (maxz-lab) [reply net/socket/transport_client_socket_unittest.cc:299]: Agreed. They should be removed.

File: starboard/tools/api_leak_detector/api_leak_detector.py [hunk 5]
```diff
@@ -94,6 +94,7 @@
     'connect',
     'clock_gettime',
     'close',
+    'fcntl',
```
Comment (y4vor) [inline starboard/tools/api_leak_detector/api_leak_detector.py:97]: I didn't see this one in the exported symbols. Do you need it?

**Comment by jellefoks** [comment]
Can you first make sure all the tests pass?

File: net/socket/udp_socket_starboard.h [hunk 6]
```diff
@@ -42,6 +42,8 @@
 
 namespace net {
 
+#if SB_API_VERSION <= 15
```
Comment (madhurajayaraman) [inline net/socket/udp_socket_starboard.h:45]: Can be in a followup, but trunk only supports SB17 now so if a file is gated behind any API version < 17, we can delete it.

**Review by y4vor — APPROVED** [review]

File: net/socket/tcp_socket_posix.cc [hunk 7]
```diff
@@ -237,8 +239,12 @@ int TCPSocketPosix::Bind(const IPEndPoint& address) {
   SockaddrStorage storage;
   if (!address.ToSockAddr(storage.addr, &storage.addr_len))
     return ERR_ADDRESS_INVALID;
-
-  return socket_->Bind(storage);
+  int rv = socket_->Bind(storage);
+  if (rv != 0){
+    DLOG(ERROR) << "TCPSocketPosix::Bind returned an error" << rv;
+    return MapSystemError(errno);
```
Comment (jellefoks) [inline net/socket/tcp_socket_posix.cc:245]: Why do we need to make this translation call? Shouldn't `Bind` just return the right type/value?

File: net/socket/tcp_socket_posix.cc [hunk 8]
```diff
@@ -251,9 +257,12 @@ int TCPSocketPosix::Accept(std::unique_ptr<TCPSocketPosix>* tcp_socket,
                            CompletionOnceCallback callback) {
   DCHECK(tcp_socket);
   DCHECK(!callback.is_null());
-  DCHECK(socket_);
   DCHECK(!accept_socket_);
 
+  if ((!socket_)) {
```
[the remark this answers is not in the document]
Comment (maxz-lab) [reply net/socket/tcp_socket_posix.cc:257]: It is needed when the socket_ isn't properly populated. This case happens when IPv6 is not supported by host.

File: net/socket/tcp_socket_posix.cc (hunk 7 above)
Comment (maxz-lab) [reply net/socket/tcp_socket_posix.cc]: You are right, it should be removed. And it is removed.

File: net/socket/udp_socket_posix.cc [hunk 9]
```diff
@@ -779,31 +779,23 @@ int UDPSocketPosix::InternalRecvFromNonConnectedSocket(IOBuffer* buf,
...
-  };
-  struct msghdr msg = {
-      .msg_name = storage.addr,
-      .msg_namelen = storage.addr_len,
-      .msg_iov = &iov,
-      .msg_iovlen = 1,
-  };
-  int result;
-  int bytes_transferred = HANDLE_EINTR(recvmsg(socket_, &msg, 0));
+  int result = -1;
```
Comment (jellefoks) [inline net/socket/udp_socket_posix.cc:782]: There are a few reasons why changing this to be different than upstream is scary. We really should support `recvmsg` in Starboard.

Comment (maxz-lab) [reply net/socket/udp_socket_posix.cc:782]: recvmsg() behaves differently from recvfrom() only when the socket is in blocking mode. All of the sockets created here are set to non-blocking by default, these two functions essentially provide the same functionality in our case.

File: net/socket/socket_posix.h [hunk 10]
```diff
@@ -19,13 +19,15 @@
...
+#if SB_API_VERSION >= 16
+
 class IOBuffer;
 struct SockaddrStorage;
 
 // Socket class to provide asynchronous read/write operations on top of the
 // posix socket api. It supports AF_INET, AF_INET6, and AF_UNIX addresses.
 class NET_EXPORT_PRIVATE SocketPosix
-    : public base::MessagePumpForIO::FdWatcher {
+    : public base::MessagePumpForIO::Watcher {
```
Comment (jellefoks) [inline net/socket/socket_posix.h:30]: Instead of this, we should remove the use of `message_pump_io_starboard.h` in  base/message_loop/message_pump_for_io.h and from there use the `base/message_loop/message_pump_libevent.h` under the  `BUILDFLAG(IS_POSIX)`, then we don't have to make these scary changes here.

If that is nontrivial and these changes are temporary, then `MessagePumpIOStarboard` should be made to have an interface compatible with `MessagePumpLibevent` so that the changes here are not necessary.

File: net/BUILD.gn [hunk 11]
```diff
@@ -1175,6 +1175,16 @@ component("net") {
       "socket/tcp_socket_starboard.h",
       "socket/udp_socket_starboard.cc",
       "socket/udp_socket_starboard.h",
+      # Use POSIX socket for SB16
```
Comment (jellefoks) [inline net/BUILD.gn:1178]: Why not filter these lists using `sb_api_version`, then you don't need to modify the source code with `#if SB_API_VERSION >= 16` and friends?

File: net/socket/udp_socket_posix.cc (hunk 9 above)
Comment (jellefoks) [reply net/socket/udp_socket_posix.cc:782]: I'm pretty sure that the Chrome implementation is correct, so any changes made should not be here.

They are not the same. There are edge case differences (wrt truncated packages) and performance differences that matter. We should avoid changing this without good reason. Perhaps the issue is that we shouldn't be setting our UDP sockets to nonblocking.

File: net/socket/socket_posix.h (hunk 10 above)
Comment (maxz-lab) [reply net/socket/socket_posix.h:30]: This CL is knid of temporary and we will decide later based on further performance analysis. 
b/302741384 tracks separate effort to migrate /consolidate different message_pumps.

File: net/BUILD.gn (hunk 11 above)
Comment (maxz-lab) [reply net/BUILD.gn:1178]: There is a chance it will be merged back to Cobalt25, and it looks more consistent with other source files when checking on SB_API versions.

File: net/socket/udp_socket_posix.cc (hunk 9 above)
Comment (jellefoks) [reply net/socket/udp_socket_posix.cc:782]: The crux is fragility in relying on the return value to detect truncated packets (in addition to understandable sensitivity in Chrome to any changes here without detailed performance investigation).

Posix says `Upon successful completion, recvfrom() shall return the length of the message in bytes` while the linux man page says that plus `These calls return the number of bytes received`. 

Those two statements contradict each other for truncated packets, since in that case the message was longer than the number of bytes received, since "the excess is discarded" as allowed in both documentation of both Posix and the linux man page.

The difference is in interpretation: Whether a truncated message is treated as a successful completion or not, and if treated as successful, whether a discarded byte is still counted as part of "the length of the message".

There is no such confusion when using `recvmsg`.

This code in upstream used to use `recvfrom` before detection of packet truncation was added in this CL: https://chromium-review.googlesource.com/c/chromium/src/+/657780

File: net/socket/socket_posix.h (hunk 10 above)
Comment (jellefoks) [reply net/socket/socket_posix.h:30]: This is clearly upstream code. Any non-upstreamed changes made for starboardization should be clearly visible as such and these changes don't, but they do actually have an impact on both the behavior and performance of this code.

The top #3 native crash in the latest releases are in a call to `ClearWatcherIfOperationsNotPending` that currently only exists in `TCPSocketStarboard`, and this PR replaces calls to `StopWatchingFileDescriptor();` calls with calls to a very similar `ClearWatcherIfOperationsNotPending` that it adds here to socket_posix. That's not a good sign.

File: net/socket/udp_socket_posix.cc (hunk 9 above)
Comment (jellefoks) [reply net/socket/udp_socket_posix.cc:782]: Also, we're going to need this to be using recvmsg for the best behavior on the network because the ECN bits aren't signaled when using recvfrom. Support for this is actively being worked on in upstream.

https://crbug.com/40285166
https://en.wikipedia.org/wiki/Explicit_Congestion_Notification

File: net/BUILD.gn (hunk 11 above)
Comment (jellefoks) [reply net/BUILD.gn:1178]: If you have seen that the same usage exists where entire files are only used for certain starboard versions that could be filtered in the gn config, then please file a bug that we can clean those and this instance here up as a followup. Minimizing changes in code we take from upstream makes rebasing easier.

File: net/socket/udp_socket_posix.cc (hunk 9 above)
Comment (maxz-lab) [reply net/socket/udp_socket_posix.cc:782]: recvmsg is restored.

**Review by yellowdoge — APPROVED** [review]
Tests passing - RS LGTM!

**Review by johnxwork — APPROVED** [review]

File: net/socket/udp_socket_posix.cc (hunk 9 above)
Comment (jellefoks) [reply net/socket/udp_socket_posix.cc:782]: The PR was merged, but this conversation was ignored?

File: net/socket/socket_posix.h (hunk 10 above)
Comment (jellefoks) [reply net/socket/socket_posix.h:30]: The PR was merged, but this conversation was ignored?

File: net/socket/tcp_socket_posix.cc [hunk 12]
```diff
@@ -237,7 +239,6 @@ int TCPSocketPosix::Bind(const IPEndPoint& address) {
   SockaddrStorage storage;
   if (!address.ToSockAddr(storage.addr, &storage.addr_len))
     return ERR_ADDRESS_INVALID;
-
```
Comment (jellefoks) [inline net/socket/tcp_socket_posix.cc:240]: Why did you delete this line in upstream code?

File: net/socket/tcp_socket_posix.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (jellefoks) [reply net/socket/tcp_socket_posix.cc:257]: Instead of changing the upstream code to handle this, we should fix the place where `Accept()` is called with `nullptr` for `tcp_socket`.

[the remark this answers is not in the document]
Comment (jellefoks) [reply net/socket/tcp_socket_posix.cc:257]: Instead of changing the upstream code to handle this, we should fix the place where `Accept()` is called with `nullptr` for `tcp_socket`.

File: net/socket/transport_client_socket_unittest.cc
Comment (jellefoks) [inline net/socket/transport_client_socket_unittest.cc:1]: Why was this merged with whitespace changes to upstream code?

File: net/proxy_resolution/pac_file_fetcher_impl_unittest.cc [hunk 13]
```diff
@@ -171,7 +171,7 @@ TEST_F(PacFileFetcherImplTest, HttpMimeType) {
                                     TRAFFIC_ANNOTATION_FOR_TESTS);
     EXPECT_THAT(result, IsError(ERR_IO_PENDING));
     EXPECT_THAT(callback.WaitForResult(), IsOk());
-    EXPECT_EQ(u"-pac.txt-\n", text);
```
Comment (jellefoks) [inline net/proxy_resolution/pac_file_fetcher_impl_unittest.cc:174]: Are you sure that it's OK that when we run it, the newline is dropped?

Comment (jellefoks) [reply net/proxy_resolution/pac_file_fetcher_impl_unittest.cc:174]: Specifically, the file being fetched 'pac.txt' does have the newline, and the upstream test has always assumed that the newline is also read, so I don't think the test was wrong.

File: net/socket/udp_socket_posix.cc (hunk 9 above)
Comment (maxz-lab) [reply net/socket/udp_socket_posix.cc:782]: I have restored recvmsg and forgot to push the change, as I explained in the chat channel. The new patch has been submitted.
